config/v1/types_infrastructure.go

```diff
@@ -573,6 +583,41 @@ type EquinixMetalPlatformStatus struct {
 	IngressIP string `json:"ingressIP,omitempty"`
 }
 
+// PowervsServiceEndpoint store the configuration of a custom url to
+// override existing defaults of PowerVS Services.
+type PowerVSServiceEndpoint struct {
+	// name is the name of the Power VS service.
+	//
+	// +kubebuilder:validation:Pattern=`^[a-z0-9-]+$`
+	Name string `json:"name"`
+
+	// url is fully qualified URI with scheme https, that overrides the default generated
+	// endpoint for a client.
+	// This must be provided and cannot be empty.
+	//
+	// +kubebuilder:validation:Pattern=`^https://`
+	URL string `json:"url"`
```

Comment:
there is a format for urls.

Is it optional? How about omitempty?

Reply:
even with this pattern it can be undefined.

Reply:
same here as well, can `+kubebuilder:validation:Required` solve the issue?

Reply:
I tried looking at here for the format https://book.kubebuilder.io/reference/markers/crd-validation.html, but didn't find anything useful, any useful reference for the url validation?

Reply:
> same here as well, can +kubebuilder:validation:Required solve the issue?

yes

Reply:
compare https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types_jsonschema.go#L53

Reply:
```go
// +kubebuilder:validation:Required
// +kubebuilder:validation:Type=string
// +kubebuilder:validation:Format=uri
```

hope this works ^^^

Reply:
so `telnet://1.2.3.4` is valid?

Reply:
> `// +kubebuilder:validation:Type=string` is implicit through the golang type.

okay

Reply:
> so `telnet://1.2.3.4` is valid?

not really,! got a point. Do think having both uri and pattern matching will help to block this?

Reply:
technically it will. Whether you want your API to be restricted to https is your choice :)

Reply:
hah, yes thanks for catching that. will update